### ISSUE TITLE
CDB_Overviews(reloid REGCLASS) fails for org users

### ISSUE THREAD

**Issue by rafatower** [CartoDB/cartodb-postgresql#218]
```
=> SELECT cartodb.CDB_DropOverviews('rtorre.yellow_tripdata_2015_pickups'::regclass);
ERROR:  relation "_vovw_19_yellow_tripdata_2015_pickups" does not exist
CONTEXT:  SQL function "cdb_overviews" statement 1
PL/pgSQL function cdb_dropoverviews(regclass) line 5 at FOR over SELECT rows
```

```
=> \dt rtorre.*yellow*                                                                                                                                                                                                                  List of relations
 Schema |                 Name                  | Type  |                       Owner                       
--------+---------------------------------------+-------+---------------------------------------------------
 rtorre | _vovw_0_yellow_tripdata_2015_pickups  | table | cartodb_user_uuid-lalala
 rtorre | _vovw_10_yellow_tripdata_2015_pickups | table | cartodb_user_uuid-lalala
 rtorre | _vovw_11_yellow_tripdata_2015_pickups | table | cartodb_user_uuid-lalala
 rtorre | _vovw_12_yellow_tripdata_2015_pickups | table | cartodb_user_uuid-lalala
 rtorre | _vovw_13_yellow_tripdata_2015_pickups | table | cartodb_user_uuid-lalala
 rtorre | _vovw_14_yellow_tripdata_2015_pickups | table | cartodb_user_uuid-lalala
 rtorre | _vovw_15_yellow_tripdata_2015_pickups | table | cartodb_user_uuid-lalala
 rtorre | _vovw_16_yellow_tripdata_2015_pickups | table | cartodb_user_uuid-lalala
 rtorre | _vovw_17_yellow_tripdata_2015_pickups | table | cartodb_user_uuid-lalala
 rtorre | _vovw_18_yellow_tripdata_2015_pickups | table | cartodb_user_uuid-lalala
 rtorre | _vovw_19_yellow_tripdata_2015_pickups | table | cartodb_user_uuid-lalala
 rtorre | _vovw_1_yellow_tripdata_2015_pickups  | table | cartodb_user_uuid-lalala
 rtorre | _vovw_2_yellow_tripdata_2015_pickups  | table | cartodb_user_uuid-lalala
 rtorre | _vovw_3_yellow_tripdata_2015_pickups  | table | cartodb_user_uuid-lalala
 rtorre | _vovw_4_yellow_tripdata_2015_pickups  | table | cartodb_user_uuid-lalala
 rtorre | _vovw_5_yellow_tripdata_2015_pickups  | table | cartodb_user_uuid-lalala
 rtorre | _vovw_6_yellow_tripdata_2015_pickups  | table | cartodb_user_uuid-lalala
 rtorre | _vovw_7_yellow_tripdata_2015_pickups  | table | cartodb_user_uuid-lalala
 rtorre | _vovw_8_yellow_tripdata_2015_pickups  | table | cartodb_user_uuid-lalala
 rtorre | _vovw_9_yellow_tripdata_2015_pickups  | table | cartodb_user_uuid-lalala
 rtorre | yellow_tripdata_2015_pickups          | table | cartodb_user_uuid-lalala
(21 rows)
```

```
=> show search_path;                                                                                                                                                                                   search_path             
------------------------------------
 "$user", public, cartodb, topology
(1 row)
```

**Comment by rafatower** [comment]
Actually this is the problem:
```
=> select * from cdb_overviews('rtorre.yellow_tripdata_2015_pickups'::regclass);
ERROR:  relation "_vovw_19_yellow_tripdata_2015_pickups" does not exist
CONTEXT:  SQL function "cdb_overviews" statement 1
```

**Comment by rafatower** [comment]
known problem: https://github.com/CartoDB/cartodb-postgresql/blob/aff7ae3e2e155de00958a8bbc7b95203ba9ffb8e/scripts-available/CDB_Overviews.sql#L103

**Comment by rafatower** [comment]
the problem is the `cdb_usertables::regclass AS overview_table` sub-expression in the select

**Comment by rafatower** [comment]
I wouldn't use CDB_UserTables cause the implementation looks faulty to me: https://github.com/CartoDB/cartodb-postgresql/blob/aff7ae3e2e155de00958a8bbc7b95203ba9ffb8e/scripts-available/CDB_UserTables.sql

(getting just the name of the table (with no schema) and casting to an infamous `name` type, which is a tricky one).

See http://www.postgresql.org/docs/current/static/datatype-character.html

**Comment by jgoizueta** [comment]
Apart from the questionable use of CDB_UserTables, the problem can be fixed with:

```
CREATE OR REPLACE FUNCTION CDBX_Overviews(reloid REGCLASS)
RETURNS TABLE(base_table REGCLASS, z integer, overview_table REGCLASS)
AS $$
  DECLARE
    schema_name TEXT;
    table_name TEXT;
  BEGIN
    SELECT * FROM cartodb._cdb_split_table_name(reloid) INTO schema_name, table_name;
    RETURN QUERY SELECT
      reloid AS base_table,
      cartodb._CDB_OverviewTableZ(cdb_usertables) AS z,
      (schema_name||'.'||cdb_usertables)::regclass AS overview_table
      FROM cartodb.CDB_UserTables()
      WHERE cartodb._CDB_IsOverviewTableOf((SELECT relname FROM pg_class WHERE oid=reloid), cdb_usertables)
      ORDER BY z;
  END
$$ LANGUAGE PLPGSQL;
```

**Comment by rafatower** [comment]
the shortest line to reproduce the problem as it's happening:
```
SELECT '_vovw_19_yellow_tripdata_2015_pickups'::name::regclass AS overview_table;
```

buuut, the trick is that from other places the search path is set to something like this (e.g: when querying from the sql api:
```
search_path: "rtorre, cartodb, cdb_dataservices_client, public"
```

and actually the query does not fail from the sql api, but it returns no rows.

**Comment by jgoizueta** [comment]
I'm reopening this because `CDB_Overviews` can still fail for organization users under some circumstances:

* Several organisation members have a table with the same name and it has overviews
* Not all the tables have the same overview levels
* We execute `CDB_Overviews` with postgres user (i.e. with privileges to access the tables from the various users)

In that case `CDB_overviews` will fail for some of the users's tables (when trying to cast to regclass the name of an overview table that the owner of the table doen't have).

To solve this the use of `CDB_UserTables` should be replaced.

The problem is not urgent, as the mentioned conditions are not met under the normal use of CDB_Overviews from the tiler or the editor.

**Comment by jgoizueta** [comment]
Solved in #224, Version 0.15.1 of the extension includes the fix.
